daemon/job/snapjob.go

```diff
@@ -0,0 +1,379 @@
+package job
+
+import (
+	"context"
+	"github.com/pkg/errors"
+//	"github.com/problame/go-streamrpc"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/zrepl/zrepl/config"
+	"github.com/zrepl/zrepl/daemon/filters"
+	"github.com/zrepl/zrepl/daemon/job/reset"
+	"github.com/zrepl/zrepl/daemon/job/wakeup"
+	"github.com/zrepl/zrepl/daemon/logging"
+	"github.com/zrepl/zrepl/daemon/pruner"
+	//"github.com/zrepl/zrepl/pruning"
+	"github.com/zrepl/zrepl/daemon/snapper"
+	"github.com/zrepl/zrepl/daemon/transport/connecter"
+	"github.com/zrepl/zrepl/endpoint"
+//	"github.com/zrepl/zrepl/replication"
+	"github.com/zrepl/zrepl/util/envconst"
+	"github.com/zrepl/zrepl/zfs"
+	"sync"
+	"time"
+)
+
+type snap_ActiveSide struct {
+	mode          snap_activeMode
+	name          string
+	clientFactory *connecter.ClientFactory
+
+	prunerFactory *pruner.SinglePrunerFactory
+
+	//promRepStateSecs *prometheus.HistogramVec // labels: state
+	promPruneSecs *prometheus.HistogramVec // labels: prune_side
+	//promBytesReplicated *prometheus.CounterVec // labels: filesystem
+
+	tasksMtx sync.Mutex
+	tasks    snap_activeSideTasks
+}
+
+
+type snap_activeSideTasks struct {
+	state ActiveSideState
+
+	// valid for state ActiveSideReplicating, ActiveSidePruneSender, ActiveSidePruneReceiver, ActiveSideDone
+	//replication *replication.Replication
+	//replicationCancel context.CancelFunc
+
+	// valid for state ActiveSidePruneSender, ActiveSidePruneReceiver, ActiveSideDone
+	//prunerSender, prunerReceiver *pruner.Pruner
+	pruner *pruner.Pruner
+
+	// valid for state ActiveSidePruneReceiver, ActiveSideDone
+	prunerCancel context.CancelFunc
+}
+
+func (a *snap_ActiveSide) Name() string { return a.name }
+
+func (a *snap_ActiveSide) GetPruner(ctx context.Context, sender *endpoint.Sender) (*pruner.Pruner) {
+    //p := &pruner.Pruner{ args: pruner.args{ctx, WithLogger(ctx, GetLogger(ctx).WithField("prune_side", "sender")), sender, sender, a.rules, envconst.Duration("ZREPL_PRUNER_RETRY_INTERVAL", 10 * time.Second), true, a.promPruneSecs.WithLabelValues("sender")}, state: pruner.State.Plan,}
+    p := a.prunerFactory.BuildSinglePruner(ctx,sender,sender)
+    /*if err != nil {
+	return nil, errors.Wrap(err, "cannot build receiver pruning rules")
+    }*/
+    return p
+}
+
+
+func (a *snap_ActiveSide) updateTasks(u func(*snap_activeSideTasks)) snap_activeSideTasks {
+	a.tasksMtx.Lock()
+	defer a.tasksMtx.Unlock()
+	var copy snap_activeSideTasks
+	copy = a.tasks
+	if u == nil {
+		return copy
+	}
+	u(&copy)
+	a.tasks = copy
+	return copy
+}
+
+
+type snap_activeMode interface {
```

Comment:
a `mode` struct is not necessary for this kind of job, since there are no modes to a snap-job.

We needed `mode` for active and passive side, where `mode` has the role of a concrete subclass, or the *strategy design pattern* if that means anything to you ;)

please remove the `mode` struct and make `mode`'s member functions functions of the job `snap_ActiveSide struct`